File: vsa/examples/01-todo-list-ts/src/contexts/tasks/delete-task/DeleteTaskHandler.ts
import { DeleteTaskCommand } from './DeleteTaskCommand';
import { TaskDeletedEvent } from './TaskDeletedEvent';
import { EventStore } from '../../../infrastructure/EventStore';
import { TaskAggregate } from '../create-task/TaskAggregate';

/**
 * Handler for DeleteTask command
 * 
 * Business Rules:
 * - Task must exist
 * - Task must not be already deleted
 */
export class DeleteTaskHandler {
  constructor(private eventStore: EventStore) {}

  async handle(command: DeleteTaskCommand): Promise<void> {
    // Load task aggregate from events
    const events = await this.eventStore.getEvents(command.id);
    
    if (events.length === 0) {
      throw new Error(`Task with ID ${command.id} not found`);
    }

    // Rebuild aggregate state
    const aggregate = new TaskAggregate();
    for (const event of events) {
      this.applyEvent(aggregate, event);
    }

    // Business rule validation
    if (aggregate.isDeleted()) {
      throw new Error('Task is already deleted');
    }

    // Create event
    const event: TaskDeletedEvent = {
      id: command.id,
      deletedAt: new Date(),
    };

    // Store event
    await this.eventStore.appendEvent(command.id, 'TaskDeleted', event);
  }

  private applyEvent(aggregate: TaskAggregate, event: any): void {
    switch (event.type) {
      case 'TaskCreated':
        aggregate.applyTaskCreated(event.data);
        break;
      case 'TaskCompleted':
        aggregate.applyTaskCompleted(event.data);
        break;
      case 'TaskDeleted':
        aggregate.applyTaskDeleted(event.data);
        break;
    }
  }
}

File: vsa/examples/05-todo-list-py/src/contexts/tasks/delete-task/DeleteTaskHandler.py
"""Handler for DeleteTaskCommand"""

from .DeleteTaskCommand import DeleteTaskCommand
from .TaskDeletedEvent import TaskDeletedEvent
from event_sourcing import Repository

class DeleteTaskHandler:
    """Handler for DeleteTaskCommand

    This handler processes the command, applies business logic,
    creates events, and persists them to the event store.
    """

    def __init__(self, repository: Repository):
        self.repository = repository

    async def handle(self, command: DeleteTaskCommand) -> None:
        """Process the command and emit events"""
        # TODO: Add validation logic

        # Create event
        event = TaskDeletedEvent(
            id=command.id,
        )

        # TODO: Persist to event store
        # await self.event_store.append(stream_name, [event])
